client/state/data-layer/wpcom-http/README.md

```diff
@@ -0,0 +1,140 @@
+
+In contrast to a fully-generic HTTP layer this one understands that requests are being made directly to the WordPress.com API and allowances are made in the HTTP Request descriptions to acknowledge this and make working with it a smoother process; for example, authentication is automatically handled.
+
+## Why?
+
+The end-goal for this two-layer middleware approach is to make offline-use first-class in Calypso and to enable things like mobile data management and battery optimization.
+
+Let's look at how this system works from a high-level perspective.
+
+## Overview
+
+<!-- the following diagram was generated in draw.io - it can be edited by pasting in the contents of the SVG itself -->
+![](https://cldup.com/unQOzvDkjtq/vSrA37.svg)
+
+Wow, how confusing!
+Well, this isn't promising to make everything easier, but it _is_ promising to make things worth while.
+
+### Step A: Component needs to update data
+
+A React component wants to indicate that a given post ought to be marked as liked.
+It's API includes two functions for accomplishing this: `likePost( siteId, postId )` and `unlikePost( siteId, postId )`.
+It knows nothing about how those two functions work, only that they are the names it uses to accomplish its goal.
+It calls `likePost` after the user clicks on the like button.
+
+### Step B: API middleware intercepts action
+
+The WordPress.com API middleware understands what it means to like a post.
+It knows that when we do so, the server needs to also be notified so that the like actually persists.
+Further, it knows that it's possible that the server might reject the like, or some other trigger might cause the like to fail.
+
+The middleware speaks with the WordPress.com servers through HTTP requests.
+These requests have special properties: they can fail, we might want them to automatically retry upon failure, they can be batched, etc.
+Therefore, instead of coding all of this logic into the API middleware itself the middleware will issue another declarative action describing the HTTP request and the meta information and policies about that request.
+The new action carries along the information for the request such as the WordPress.com API path and associated query data, an action to dispatch if the request is a success, and an action to dispatch if the request fails.
+
+Because this middleware is merely attempting to keep the server in sync with Calypso's updated state, it needs to pass the action along so that the reducers get the chance to update the app state and reflect the change.
+
+### Step C: HTTP middleware runs actual HTTP request
+
+The HTTP middleware understands what it means to run HTTP requests.
+It may take some liberties in how it decides to actually execute the requests.
+This liberty is what allows us to provide the global policies like retry behavior, offline behavior, and batching.
```

Comment:
I wonder if this would benefit from more of a blogging style. following your comment above, we might want to pull out some snippets in block quotes to highlight the _why_s